openupgrade_scripts/scripts/microsoft_calendar/18.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,171 @@
+    """
+
+    openupgrade.add_fields(
+        env,
+        [
+            (
+                "ms_universal_event_id",
+                "calendar.recurrence",
+                "calendar_recurrence",
+                "char",
+                False,
+                "microsoft_calendar",
+                False,
+            )
+        ],
+    )
+
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE calendar_recurrence ce
+        SET ms_universal_event_id = (
+            SELECT
+                CASE
+                    WHEN array_length(string_to_array(microsoft_id, ':'), 1) >= 2
+                    THEN string_to_array(microsoft_id, ':')[2]
+                END
+            FROM calendar_recurrence
+            WHERE id = ce.id
+        );
+        """,
+    )
+
+
+def microsoft_calendar_credentials_to_res_users_settings(env):
+    """
+    microsoft.calendar.credentials has been deleted and now settings are
+    stored in res.users.settings. For each user in microsoft.calendar.credentials
+    user_ids (one2many) field we must copy into that user's res.users.settings
+    (user_id, many2one) the following fields:
+    * calendar_sync_token -> microsoft_calendar_sync_token
+    * last_sync_date -> microsoft_last_sync_date
+    * synchronization_stopped -> microsoft_synchronization_stopped
+    """
+
+    openupgrade.add_fields(
+        env,
+        [
+            (
+                "microsoft_calendar_sync_token",
+                "res.users.settings",
+                "res_users_settings",
+                "char",
+                False,
+                "microsoft_calendar",
+                False,
+            )
+        ],
+    )
+
+    openupgrade.add_fields(
+        env,
+        [
+            (
+                "microsoft_last_sync_date",
+                "res.users.settings",
+                "res_users_settings",
+                "datetime",
+                False,
+                "microsoft_calendar",
+                False,
+            )
+        ],
+    )
+
+    openupgrade.add_fields(
+        env,
+        [
+            (
+                "microsoft_last_sync_date",
+                "res.users.settings",
+                "res_users_settings",
+                "boolean",
+                False,
+                "microsoft_synchronization_stopped",
+                False,
+            )
+        ],
+    )
+
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE res_users_settings AS rus
+        SET
+            microsoft_calendar_sync_token = cred.calendar_sync_token,
+            microsoft_last_sync_date = cred.last_sync_date,
+            microsoft_synchronization_stopped = cred.synchronization_stopped
+        FROM res_users AS ru
+        JOIN microsoft_calendar_credentials AS cred
+            ON ru.microsoft_calendar_credentials_id = cred.id
+        WHERE rus.user_id = ru.id;
+        """,
+    )
+
+
+@openupgrade.migrate()
+def migrate(env, version):
```

Comment:
```suggestion
def migrate(env, version):
    openupgrade.add_columns(env, [
        ("calendar.event", "ms_universal_event_id", "char"),
        ("calendar.recurrence", "ms_universal_event_id", "char"),
        ("res.users.settings", "microsoft_calendar_sync_token", "char"),
        ("res.users.settings", "microsoft_last_sync_date", "datetime"),
        ("res.users.settings", "microsoft_synchronization_stopped", "boolean"),
    ])
```
It's better to group them here, and let other methods only fill values. Also, better create column than field.